File: core/descriptor.loader.java/src/org/jetbrains/jet/lang/resolve/java/lazy/descriptors/LazyJavaMemberScope.java
package org.jetbrains.jet.lang.resolve.java.lazy.descriptors;

import com.google.common.collect.Lists;
import com.intellij.openapi.util.Computable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.jet.lang.descriptors.*;
import org.jetbrains.jet.lang.resolve.java.JavaClassFinder;
import org.jetbrains.jet.lang.resolve.lazy.storage.NotNullLazyValue;
import org.jetbrains.jet.lang.resolve.lazy.storage.StorageManager;
import org.jetbrains.jet.lang.resolve.name.LabelName;
import org.jetbrains.jet.lang.resolve.name.Name;
import org.jetbrains.jet.lang.resolve.scopes.JetScope;

import java.util.Collection;
import java.util.Collections;

public abstract class LazyJavaMemberScope implements JetScope {
    protected final JavaClassFinder finder;
    protected final StorageManager storageManager;

    private final DeclarationDescriptor containingDeclaration;
    private final NotNullLazyValue<Collection<DeclarationDescriptor>> allDescriptors;

    protected LazyJavaMemberScope(@NotNull DeclarationDescriptor containingDeclaration, @NotNull StorageManager storageManager, @NotNull JavaClassFinder finder) {
        this.finder = finder;
        this.containingDeclaration = containingDeclaration;
        this.storageManager = storageManager;

        this.allDescriptors = storageManager.createLazyValue(new Computable<Collection<DeclarationDescriptor>>() {
            @Override
            public Collection<DeclarationDescriptor> compute() {
                return computeAllDescriptors();
            }
        });

    }

    @NotNull
    @Override
    public DeclarationDescriptor getContainingDeclaration() {
        return containingDeclaration;
    }

    @Nullable
    @Override
    public ClassDescriptor getObjectDescriptor(@NotNull Name name) {
        // No object declarations in Java
        return null;
    }

    @NotNull
    @Override
    public Collection<ClassDescriptor> getObjectDescriptors() {
        // No object declarations in Java
        return Collections.emptyList();
    }

    @Nullable
    @Override
    public NamespaceDescriptor getNamespace(@NotNull Name name) {
        // Should be resolved elsewhere
        return null;
    }

    @Nullable
    @Override
    public VariableDescriptor getLocalVariable(@NotNull Name name) {
        // Not a local scope
        return null;
    }

    @NotNull
    @Override
    public Collection<DeclarationDescriptor> getDeclarationsByLabel(@NotNull LabelName labelName) {
        // A member scope has no labels
        return Collections.emptySet();
    }

    @NotNull
    @Override
    public Collection<DeclarationDescriptor> getOwnDeclaredDescriptors() {
        return getAllDescriptors();
    }

    @NotNull
    @Override
    public Collection<DeclarationDescriptor> getAllDescriptors() {
        return allDescriptors.compute();
    }

    private Collection<DeclarationDescriptor> computeAllDescriptors() {
        Collection<DeclarationDescriptor> result = Lists.newArrayList();

        for (Name name : getAllClassNames()) {
            DeclarationDescriptor descriptor = getClassifier(name);
            assert descriptor != null : "Descriptor not found for name " + name + " in " + containingDeclaration;
            result.add(descriptor);
        }

        for (Name name : getAllFunctionNames()) {
            result.addAll(getFunctions(name));
        }

        for (Name name : getAllPropertyNames()) {
            result.addAll(getProperties(name));
        }

        addExtraDescriptors(result);
        return result;
    }

    @NotNull
    protected abstract Collection<Name> getAllClassNames();

    @NotNull
    protected abstract Collection<Name> getAllPropertyNames();

    @NotNull
    protected abstract Collection<Name> getAllFunctionNames();

    protected abstract void addExtraDescriptors(@NotNull Collection<? super DeclarationDescriptor> result);

}

File: core/descriptor.loader.java/src/org/jetbrains/jet/lang/resolve/java/lazy/descriptors/LazyJavaMemberScope.kt
package org.jetbrains.jet.lang.resolve.java.lazy.descriptors

import com.google.common.collect.Lists
import org.jetbrains.jet.lang.descriptors.*
import org.jetbrains.jet.lang.resolve.java.JavaClassFinder
import org.jetbrains.jet.lang.resolve.lazy.storage.NotNullLazyValue
import org.jetbrains.jet.lang.resolve.lazy.storage.StorageManager
import org.jetbrains.jet.lang.resolve.name.LabelName
import org.jetbrains.jet.lang.resolve.name.Name
import org.jetbrains.jet.lang.resolve.scopes.JetScope
import java.util.Collections
import org.jetbrains.jet.utils.emptyList

public abstract class LazyJavaMemberScope(
        private val _containingDeclaration: DeclarationDescriptor, 
        protected val storageManager: StorageManager, 
        protected val finder: JavaClassFinder
) : JetScope {
    private val allDescriptors: NotNullLazyValue<MutableCollection<DeclarationDescriptor>> = storageManager.createLazyValue{computeAllDescriptors()}

    override fun getContainingDeclaration() = _containingDeclaration

    // No object can be defined in Java
    override fun getObjectDescriptor(name: Name): ClassDescriptor? = null
    override fun getObjectDescriptors() = emptyList<ClassDescriptor>()

    // namespaces should be resolved elsewhere
    override fun getNamespace(name: Name): NamespaceDescriptor? = null

    override fun getLocalVariable(name: Name): VariableDescriptor? = null
    override fun getDeclarationsByLabel(labelName: LabelName) = emptyList<DeclarationDescriptor>()

    override fun getOwnDeclaredDescriptors() = getAllDescriptors()
    override fun getAllDescriptors() = allDescriptors.compute()

    private fun computeAllDescriptors(): MutableCollection<DeclarationDescriptor> {
        val result = arrayListOf<DeclarationDescriptor>()

        for (name in getAllClassNames()) {
            val descriptor = getClassifier(name)
            assert(descriptor != null) {"Descriptor not found for name " + name + " in " + getContainingDeclaration()}
            result.add(descriptor!!)
        }

        for (name in getAllFunctionNames()) {
            result.addAll(getFunctions(name))
        }

        for (name in getAllPropertyNames()) {
            result.addAll(getProperties(name))
        }

        addExtraDescriptors(result)

        return result
    }
    protected abstract fun getAllClassNames(): Collection<Name>
    protected abstract fun getAllPropertyNames(): Collection<Name>
    protected abstract fun getAllFunctionNames(): Collection<Name>
    protected abstract fun addExtraDescriptors(result: MutableCollection<in DeclarationDescriptor>)


}
